### ISSUE TITLE
Rust support

### ISSUE THREAD

**Issue by jrmuizel** [eurecom-s3/symcc#1]
This would be really nice to have.

**Comment by sebastianpoeplau** [comment]
We think so too :) There's someone currently looking into it - we'll keep you posted.

**Comment by unseddd** [comment]
@sebastianpoeplau I have also started working on this a bit, would be happy to share notes on what I have so far. Not meaning to step on anyone's toes, just excited to get Rust support implemented.

What I found so far (with help from rustc wg-llvm):

- compiling a custom rustc, with added symbolize pass in [rust/compiler/rustc_codegen_llvm/src/back/write.rs](https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_llvm/src/back/write.rs#L611)
  - requires writing a wrapper around `libSymbolize` pass
  - maybe stage 0 gets compiled with the pass added, then later stages get `library/{alloc,core,std}` instrumented?
- add the symbolize pass as a sanitizer in upstream LLVM, and use existing machinery to add the [pass via LLVM args](https://rustc-dev-guide.rust-lang.org/sanitizers.html)
  - may only require a custom `library/{alloc,core,std}` and linking, versus carrying around a full `rustc` build
  - requires no extra code in `rust` code base
  - requires merge into upstream LLVM

I'm new to LLVM code/project, so still learning how everything fits together. Appreciate any feedback/collaboration,.

Thanks for writing SymCC, and the associated paper.

**Comment by unseddd** [comment]
@sebastianpoeplau I have also started working on this a bit, would be happy to share notes on what I have so far. Not meaning to step on anyone's toes, just excited to get Rust support implemented.

What I found so far (with help from rustc wg-llvm):

- compiling a custom rustc, with added symbolize pass in [rust/compiler/rustc_codegen_llvm/src/back/write.rs](https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_llvm/src/back/write.rs#L611)
  - requires writing a wrapper around `libSymbolize` pass
  - maybe stage 0 gets compiled with the pass added, then later stages get `library/{alloc,core,std}` instrumented?
- add the symbolize pass as a sanitizer in upstream LLVM, and use existing machinery to add the [pass via LLVM args](https://rustc-dev-guide.rust-lang.org/sanitizers.html)
  - may only require a custom `library/{alloc,core,std}` and linking, versus carrying around a full `rustc` build
  - requires no extra code in `rust` code base
  - requires merge into upstream LLVM, or rust's LLVM fork

I'm new to LLVM code/project, so still learning how everything fits together. Appreciate any feedback/collaboration,.

Thanks for writing SymCC, and the associated paper.

**Comment by sebastianpoeplau** [comment]
Thanks for sharing @unseddd! In our initial work, we went for the first approach. While getting SymCC added to upstream LLVM would allow a cleaner solution and make it easier for people to use SymCC in general, a lot of continued effort would be required to keep up with LLVM's pace of development and to provide high-quality support for all its targets.

I'm trying to find the time to finish the work on Rust support from where the student left off... He was able to include the SymCC pass into the LLVM version that rustc uses, and we determined that we'd need the stage-0 compiler and the stage-1 libraries, as you said. Finally, we'd have to make sure that the new compiler links or compiles SymCC's `libSymRuntime` into binaries.

**Comment by unseddd** [comment]
> While getting SymCC added to upstream LLVM would allow a cleaner solution and make it easier for people to use SymCC in general, a lot of continued effort would be required to keep up with LLVM's pace of development and to provide high-quality support for all its targets.

I can't guarantee how long-term I can provide development and support (future being uncertain), but I want to get more involved with LLVM development. Willing to experiment with adding a symbolize pass in-tree. `llvm/lib/Transforms/Instrumentation` seems like a good place for smth like a `SymbolizePass.cpp` to live, what do you think?

I had some success with building SymCC in Rust's LLVM tree as well, but haven't figured out how to get the codegen backend to load the `libSymbolize.so` library (`rustc_codegen_llvm` looks for static `.a` libs). Still working on it, though.

> I'm trying to find the time to finish the work on Rust support from where the student left off... He was able to include the SymCC pass into the LLVM version that rustc uses, and we determined that we'd need the stage-0 compiler and the stage-1 libraries, as you said. Finally, we'd have to make sure that the new compiler links or compiles SymCC's libSymRuntime into binaries.

I have some spare cycles to throw at the problem, so I'm willing to help out. Will look into how to link `libSymRuntime` once I get a working symbolized `rustc` build. Does `libSymRuntime` need to be linked in the `rustc` build, or only for binaries being tested?

Thanks for your feedback.

**Comment by sebastianpoeplau** [comment]
> I can't guarantee how long-term I can provide development and support (future being uncertain), but I want to get more involved with LLVM development. Willing to experiment with adding a symbolize pass in-tree. `llvm/lib/Transforms/Instrumentation` seems like a good place for smth like a `SymbolizePass.cpp` to live, what do you think?

Yes, I agree. And for the custom build of `rustc` and its LLVM I think it's easiest to compile SymCC in-tree anyway, so that we avoid the challenges around locating and loading `libSymbolize.so`. We're building a custom version of the compiler already, so we can as well patch the SymCC pass in.

> I had some success with building SymCC in Rust's LLVM tree as well, but haven't figured out how to get the codegen backend to load the `libSymbolize.so` library (`rustc_codegen_llvm` looks for static `.a` libs). Still working on it, though.

Maybe it's easiest to move SymCC's `compiler` directory into the LLVM tree. The [LLVM docs](https://llvm.org/docs/CMake.html#developing-llvm-passes-out-of-source) suggest that the changes needed in the build system should be relatively minor.

> I have some spare cycles to throw at the problem, so I'm willing to help out. Will look into how to link `libSymRuntime` once I get a working symbolized `rustc` build. Does `libSymRuntime` need to be linked in the `rustc` build, or only for binaries being tested?

That's great! As for `libSymRuntime`, it's only needed to _execute_ instrumented code, not to _build_ it. So while `rustc` needs to build the runtime into the executables that it creates, I don't think it will itself need to be built against it.

**Comment by unseddd** [comment]
@sebastianpoeplau Got a working build of `rustc` with a `symbolize` pass on `rust-1.45.0`!

My working branches for [rustc](https://github.com/unseddd/rust/tree/symrs-1.45.0) and [rust-llvm](https://github.com/unseddd/llvm-project/tree/symcc-10.0).

I'll be working on instrumenting binaries, and updating `Symbolize/Symbolizer.cpp` for LLVM 11.0. The main thing to fix is calling into `IRB.CreateCall`, since the API is different in 11.0. The first argument is now a `FunctionType*` or `FunctionCallee` instead of `Value*`. Most uses are straightforward to fix, like when given an `...Inst*` type in the caller. Still not sure yet how to go from `Value*` to `FunctionType*` or `FunctionCallee` like in [Symbolizer::forceBuildRuntimeCall](https://github.com/unseddd/llvm-project/blob/symcc-10.0/llvm/lib/Transforms/Symbolize/Symbolizer.cpp#L889).

Will keep experimenting :)

**Comment by sebastianpoeplau** [comment]
Wow, that's really nice! Great progress :)

**Comment by unseddd** [comment]
> Wow, that's really nice! Great progress :)

Unfortunately, my celebration was premature :sweat_smile:

The build compiles, but the `symbolize` pass isn't registered with LLVM. So, calls to `rustc -C passes='symbolize'` are no-ops...

Working on converting the pass to use LLVM's new pass manager, to see if that helps with registry. Running into a roadblock with LLVM not linking against `libLLVMSymbolize.a`, but I'm probably just doing something dumb.

Will the legacy pass manager work with an in-tree pass? Even with putting `addSymbolizePass` + friends in `llvm/lib/Transforms/Symbolize/Pass.cpp`, the pass isn't getting registered. Am I doing something obviously wrong?

**Comment by unseddd** [comment]
Tried moving SymCC in-tree in LLVM, and using the new pass manager. Running into a linking issue, but will keep experimenting.

(Apologies for deleting previous comments, went overzealous with housekeeping)
